src/core/cgroup.c

```diff
@@ -1172,17 +1173,42 @@ static void cgroup_context_apply(
                 cgroup_apply_firewall(u);
 }
 
+CGroupMask get_cpu_accounting_mask(void) {
+        static CGroupMask needed_mask = 0;
+
+        /* On 4.15+ with unified hierarchy, cpu.stat's usage_usec is provided
+         * externally from the CPU controller, which means we don't need to
+         * enable the CPU controller just to get metrics.
+         */
+
+        if (needed_mask == 0) {
+                struct utsname u;
+                needed_mask |= CGROUP_MASK_CPUACCT;
+                assert_se(uname(&u) >= 0);
+                if (str_verscmp(u.release, "4.15") < 0 || !cg_all_unified())
+                        needed_mask |= CGROUP_MASK_CPU;
+        }
+
+        return needed_mask;
+}
+
+bool cpu_accounting_is_cheap(void) {
+        return !(get_cpu_accounting_mask() & CGROUP_MASK_CPU);
+}
+
 CGroupMask cgroup_context_get_mask(CGroupContext *c) {
         CGroupMask mask = 0;
 
         /* Figure out which controllers we need */
 
-        if (c->cpu_accounting ||
-            cgroup_context_has_cpu_weight(c) ||
+        if (cgroup_context_has_cpu_weight(c) ||
             cgroup_context_has_cpu_shares(c) ||
             c->cpu_quota_per_sec_usec != USEC_INFINITY)
                 mask |= CGROUP_MASK_CPUACCT | CGROUP_MASK_CPU;
```

Comment:
actually, we might as well start being accurate here, and only set 'CGROUP_MASK_CPU` in this case...